### PR TITLE
Check for resource availability before performing tests.

### PR DESCRIPTION
**Pull request by seman**
Added code to check for resource availability before performing cwr tests. When you add the following fields in the test plan, it automatically checks if the number of machines and CPUs are available from the provider before running the test:

```
cloud_resource:
    machines: 5
    cpus: 10

```
It checks resource availability by calculating how many resources you are already using and how many more you are requesting from the provider. In case of not having enough resources, it generates `resource not available` test result. See attached image.

I tested the code on running AWS, GCE, Azure and LXD. 

I will update the doc once this is landed.

resource-not-available

### CODE REVIEW

File: cloudweatherreport/cloudresource/resource.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    pass
+
+
+def get_credential_name(creds, cloud_name):
+    cred_name = creds.get('credentials', {}).get(cloud_name, {}).get(
+        'default-credential')
+    if cred_name:
+        return cred_name
+    # If no default credential name, get the first name from the list
+    names = list(creds.get('credentials', {}).get(cloud_name, {}).keys())
```
Comment (johnsca) [inline cloudweatherreport/cloudresource/resource.py]: My AWS has a `default-region` key that would conflict with this.  I wonder if `default-credential` and `default-region` are the only reserved keys, or if there's a list somewhere?

Also, if there are more than one credential, this could return a different one each time.  Juju's behavior when there are multiple credentials but no default is to error out; I think we should probably follow that.

File: cloudweatherreport/run.py [hunk 2]
```diff
@@ -146,35 +152,108 @@ def load_report(self, datastore, index, test_plan):
...
+        try:
+            return is_resource_available(
+                cloud=cloud, region=region,
+                num_of_instances=test_plan.cloud_resource['machines'],
+                num_of_security_groups=1,
+                num_of_cpus=test_plan.cloud_resource['cpus'],
+                instance_limit=machine_limit,
+                security_group_limit=sec_limit,
+                cpu_limit=cpu_limit,
+                credentials_name=os.getenv('{}')
```
Comment (johnsca) [inline cloudweatherreport/run.py]: This doesn't seem right.

File: cloudweatherreport/run.py [hunk 3]
```diff
@@ -146,35 +152,108 @@ def load_report(self, datastore, index, test_plan):
...
+            )
+        except (CloudNotSupported, UnknownCloudName,
+                UnknownCredentialName) as e:
```
Comment (johnsca) [inline cloudweatherreport/run.py]: I think we should treat `CloudNotSupported` differently; it's not really an error, since you're not pre-screening the list, so we shouldn't log it as one.

I also considered saying that the other errors should be more fatal, but then realized that since this is just a pre-check, it's probably better to fail softly and hope the test can pass anyway.  I do worry a bit about missing errors and not having the pre-checks run consistently.

File: cloudweatherreport/run.py [hunk 4]
```diff
@@ -146,35 +152,108 @@ def load_report(self, datastore, index, test_plan):
...
+                test_result = model.SuiteResult(
+                    provider=env.provider_name,
+                    test_outcome='Error Running Tests',
+                    tests=[model.TestResult(
+                        name='Error Running Tests',
+                        duration=0.0,
+                        output=error,
+                        result='INFRA',
+                    )])
+                return False
```
Comment (johnsca) [inline cloudweatherreport/run.py]: This `return False` isn't actually part of your change but I just noticed that this is entirely skipping writing the report.  We should ensure that the report gets written with the INFRA result.  Same for the controller connection error at the top of this method as well.  If you don't mind fixing that in this PR, great, but we can do it in another issue instead if you'd prefer.

File: cloudweatherreport/utils.py [hunk 5]
```diff
@@ -301,3 +302,9 @@ def write_to_datastore(datastore, index, update_summary=False):
     datastore.write(index.full_index_filename_html, index.as_html())
     datastore.write(index.summary_filename_json, index.summary_json())
     datastore.write(index.summary_filename_html, index.summary_html())
+
+
+def juju_cmd(cmd):
+    cmd = cmd.split() if isinstance(cmd, str) else cmd
+    cmd = ['juju'] + cmd
+    return subprocess.check_output(cmd)
```
Comment (johnsca) [inline cloudweatherreport/utils.py]: This will return `bytes` and should be decoded, no?

File: cloudweatherreport/cloudresource/resource.py (hunk 1 above)
Comment (seman) [inline cloudweatherreport/cloudresource/resource.py]: Thank you for catching this. According to this doc, I only see `default-region` and `default-credential` 
https://jujucharms.com/docs/2.0/credentials

I think you right about we should follow what Juju does. If there are multiple credential names and no default, we should raise an exception.

File: cloudweatherreport/utils.py (hunk 5 above)
Comment (seman) [inline cloudweatherreport/utils.py]: Yes, we should if we need to move to Python 3 in future.

File: cloudweatherreport/run.py (hunk 4 above)
Comment (seman) [inline cloudweatherreport/run.py]: I saw that too and I was thinking to do in another PR.

File: cloudweatherreport/run.py (hunk 2 above)
Comment (seman) [inline cloudweatherreport/run.py]: Yep, will update it.

File: cloudweatherreport/run.py (hunk 4 above)
Comment (johnsca) [inline cloudweatherreport/run.py]: Created #104 to track.

File: cloudweatherreport/utils.py (hunk 5 above)
Comment (johnsca) [inline cloudweatherreport/utils.py]: Oh, I forgot that this is still py2.  Disregard that for now, then.

**Comment by seman** [comment]
- Raise an exception when there are more than one credential names and no default.
- Skip the reserved key `default-region` in credentials list. 
- Log info when `CloudNotSupported` is raised.
- Updated unit test.
